Add main configuration API reference pages

kylegach opened the pull request:
Linked to #17243 

Related discussion: https://github.com/storybookjs/storybook/discussions/22480#discussioncomment-5852069

<!-- Thank you for contributing to Storybook! If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

## What I did

- Add an overview page for the main configuration object
- Add a page for each property of the main configuration object (e.g. `framework`, `stories`, `addons`, etc.)
- Rename TOC item @storybook/blocks → Doc Blocks, for consistency
- Update snippets & guides related to main configuration API reference pages

![Screenshot of main configuration overview page](https://github.com/storybookjs/storybook/assets/486540/e8e8598c-9648-451e-8752-192666919b8e)

## How to test

1. Follow the steps in the [contributing instructions](https://storybook.js.org/docs/react/contribute/new-snippets#preview-your-work) for this branch, `api-reference-main-config`
    - There's a lot of snippets referenced, so this step is critical!
2. Check for:
    a. Completeness
        - Is every property documented fully?
    b. Correctness
        - Types
        - Required or not
        - Default values
        - Descriptions
        - Example snippets
    c. Consistency in structure

## Checklist

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to make sure your PR is ready to be reviewed. -->

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [x] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->

kylegach committed:
Rename TOC item: @storybook/blocks -> Doc blocks

- Always sort required properties first

kylegach committed:
Add /docs/api/main-config-* pages

- Add relevant snippets

kylegach committed:
Update snippets & guides related to main-config-* pages

kylegach commented on docs/api/main-config-babel.md:
```diff
@@ -0,0 +1,102 @@
+title: 'config.babel'
+---
+
+<!-- TODO: Is `babel` or `babelDefault` the one we prefer? -->
```
I found both `babel` and `babelDefault` (with identical types) in the types. Which is preferred? Do we need to document both?

ndelangen replied:
Yeah addon creators should use `babelDefault` and `main.ts` can simply use `babel`, to ensure the user has a final say.

kylegach commented on docs/api/main-config-babel.md:
```diff
@@ -0,0 +1,102 @@
+https://github.com/storybookjs/storybook/blob/9d495a7c645331a798f67f7c548dec6ea4e1820f/code/lib/types/src/modules/core-common.ts#L61-L75
+
+So is the `FileSystemCache` type, although that belongs to a third-party library we can reference: `file-system-cache`
+-->
```
I copy/pasted the above from [`/code/lib/types/src/modules/core-common.ts`](https://github.com/storybookjs/storybook/blob/9d495a7c645331a798f67f7c548dec6ea4e1820f/code/lib/types/src/modules/core-common.ts). Is this needed and useful?

kylegach commented on docs/api/main-config-config.md:
```diff
@@ -0,0 +1,9 @@
+
+Type: `string[] | ((config: string[], options: Options) => string[] | Promise<string[]>)`
+
+Deprecated: `true`
```
While I don't think the guides or tutorials should be obligated to include deprecated features, I think the API references should.

kylegach commented on docs/api/main-config-core.md:
```diff
@@ -0,0 +1,106 @@
+
+<!-- prettier-ignore-end -->
+
+## `config.core.builder`
```
I'm not sure about including the full object path in the header like this. I do the same in the title, and in the list of properties on the Overview page.

I tried without it, e.g. the title of this page would `core` and this heading would be `builder`, but it seemed to lack context.

Curious for others' thoughts.

kylegach replied:
I decided to simplify the headings to only include the property name itself, e.g. this one is now `builder`.

To help with context, I added this between the page title (`core`, in this example) and the `Type`:

```
Parent: [main.js|ts configuration](./Overview.md)
```

kylegach commented on docs/api/main-config-core.md:
```diff
@@ -0,0 +1,106 @@
+
+Type: `ChannelOptions`
+
+<!-- TODO: No idea what this is for? -->
```
I need help describing what this is for, please.

kylegach commented on docs/api/main-config-preview-annotations.md:
````diff
@@ -0,0 +1,24 @@
+</div>
+
+```ts
+// @storybook/nextjs framework's src/preset.ts
````
This is the only example where it's not part of `main.ts`, because it didn't seem like it would be used that way. Is that correct?

kylegach commented on docs/api/main-config-preview-main-template.md:
```diff
@@ -0,0 +1,105 @@
+
+Default: See example below
+
+Override the preview's main page template, with a reference to a file containing an `.ejs` template that will be interpolated with environment variables.
```
> interpolated with environment variables

This part felt incomplete without including the default template, but that's now a maintenance burden. Any concerns?

kylegach commented on docs/api/main-config.md:
```diff
@@ -0,0 +1,42 @@
+- [`config.typescript`](./main-config-typescript.md)
+- [`config.viteFinal`](./main-config-vite-final.md)
+- [`config.webpackFinal`](./main-config-webpack-final.md)
+- [`config.config`](./main-config-config.md) (deprecated)
```
Should I include the experimental [`config.storyIndexers`](https://github.com/storybookjs/storybook/blob/next/code/lib/types/src/modules/core-common.ts#L375) here (it would go immediately above `config.config`)?

shilman replied:
No. We're probably going to change the API in 7.1 so that would be the best time to document it

kylegach commented on docs/configure/typescript.md:
```diff
@@ -6,34 +6,19 @@ Storybook provides an integrated [TypeScript](https://www.typescriptlang.org/) e
-
 <!-- prettier-ignore-end -->
 
-| Configuration element | Description                                                                                                                                                                                              |
```
This entire table is now better documented by the API reference. Not entirely sure why it was in the `configure/typescript` guide to begin with, honestly.

jonniebigodes replied:
That table was leftover from the updates to the TypeScript documentation for 7.0 as to give an overview of the options available, as cannot ascertain if the reader would check it on the overview page or not. But now that the API documentation is up, I'm good with removing it

kylegach commented on docs/snippets/common/main-config-addons.ts.mdx:
```diff
@@ -0,0 +1,22 @@
+import type { StorybookConfig } from '@storybook/your-framework';
+
+const config: StorybookConfig = {
+  framework: '@storybook/vue3-vite',
```
I wonder if we should use `'@storybook/your-framework'` here too, to match the comment above?

jonniebigodes replied:
@kylegach we should, If you could update the examples in your pull request and I'll pick on them and update any other references that are still around. Sounds good?

kylegach commented on docs/snippets/common/storybook-main-add-sass-config.js.mdx:
```diff
@@ -3,18 +3,13 @@
-    '@storybook/addon-essentials',
-    '@storybook/addon-interactions',
-  ],
   webpackFinal: async (config, { configType }) => {
```
This is used as the example for `config.webpackFinal`, but I believe this particular config is now covered by addon-styling. Is there a better example to use?

kylegach replied:
Going to replace this with a path aliasing example.

kylegach commented on docs/toc.js:
```diff
@@ -445,7 +445,114 @@ module.exports = {
+          title: 'Main configuration',
+          pathSegment: '',
+          type: 'menu',
+          children: [
```
These (along with _every_ list of anything within the API references) are ordered like so:

1. Overview
2. Required properties
3. Optional properties
4. Experimental properties
5. Deprecated properties

And each of those are then ordered alphabetically.

domyen commented:
For people googling for "main.js config" or "main.ts config", I don't think the keywords in the headings would come through as obvious. 

Perhaps calling that out in a separate section would help:
- Add an h2 before the snippet that says "main.js or main.ts" 
- Below that section, pull the description of "Configuration for Storybook is defined in `.storybook/main.js|ts`, which is located relative to the root of your project. A typical..."

In the sidebar:
- "Main configuration" » "main.js|ts configuration" or just "main.js|ts"

kylegach commented:
@domyen —

> For people googling for "main.js config" or "main.ts config", I don't think the keywords in the headings would come through as obvious.

Here's some my thinking on the H1:

"main.js|ts configuration" is the most correct/expected, but looks terrible because of the `|`. That pipe also prevents a search for `main.ts` from matching in Algolia (and probably Google).

![](https://github.com/storybookjs/storybook/assets/486540/5dc1929f-7880-4143-8fae-b955626b5a40)

"main.js configuration" or "main.ts configuration" too heavily implies that only one of the file extensions is accepted and the one not-used is then not searchable.

Thus, the helpfully ambiguous "Main configuration".

> Perhaps calling that out in a separate section would help:
> 
> Add an h2 before the snippet that says "main.js or main.ts"
> Below that section, pull the description of "Configuration for Storybook is defined in .storybook/main.js|ts, which is located relative to the root of your project. A typical..."

Y'know, I thought that heading might look a little strange. It does, imo, but not nearly as much as I feared. I'll commit that.

![](https://github.com/storybookjs/storybook/assets/486540/7c413c48-0e46-49ec-8ac2-b2a2e10aff8d)

> In the sidebar:
> 
> * "Main configuration" » "main.js|ts configuration" or just "main.js|ts"

I tried that:

![](https://github.com/storybookjs/storybook/assets/486540/14fff946-de6c-4a0a-aa69-11468f46a8b5)

The `|` in our font makes it hard to read, imo.

kylegach committed:
Update Overview to ensure "main.js" and "main.ts" are searchable

domyen commented:
<img width="211" alt="image" src="https://github.com/storybookjs/storybook/assets/263385/c64da0f8-bd01-4dd4-9a17-b14d28d5c6d6">

For the sidebar, how about "Main configuration" to "main.js or main.ts"

kylegach commented:
@domyen —

> For the sidebar, how about "Main configuration" to "main.js or main.ts"

![](https://github.com/storybookjs/storybook/assets/486540/0643b012-5f66-418c-b26c-aec6fc2c853f)

I still think "configuration" should be there.

domyen commented:
Thanks for sharing that, I think the `main.js|ts configuration` is probably the most straightforward to me. Without the `.js|ts` I would not know it's a file. 

wdty @shilman

kylegach committed:
Simplify presentation of properties

- # config.core -> # core
- ## `config.core.builder` -> ## `builder`
- Add "Parent: [main.js|ts configuration](./Overview.md)"
- Update TOC menu title

kylegach committed:
Update main.js|ts snippets for consistency

- Use `main-config-<property>-description` format, if possible
- Use "your-framework" wherever possible
- Always include `framework` & `stories` properties
- Remove unnecessary properties

kylegach committed:
Add examples for all main config sub properties

kylegach commented on docs/api/main-config-typescript.md:
```diff
@@ -0,0 +1,105 @@
+
+Type: `'react-docgen' | 'react-docgen-typescript' | false`
+
+Default:
+
+- `false`: if `@storybook/react` is not installed
+- `'react-docgen-typescript'`: if `@storybook/react` and `typescript` are installed
+- `'react-docgen'`: if `@storybook/react` is installed
```
Is this correct?

kylegach commented on docs/api/main-config-vite-final.md:
```diff
@@ -0,0 +1,26 @@
+
+## `Options`
+
+<!-- TODO: Is it alright to reference `babel`'s Options here? -->
```
Just highlighting this TODO.

shilman reviewed (commented):
Amazing work @kylegach & sorry for the slow review. Commented a bit, but it's a huge PR so probably more to discuss.

kylegach commented:
Related discussion: https://github.com/storybookjs/storybook/discussions/22629#discussioncomment-5941678

kylegach committed:
Address comments:

- Add `babelDefault` and `managerHead` pages
- Fix broken links
- Clarify when some properties are primarily used by addon authors
- Document simplified options for `babel`, `babelDefault`, `viteFinal`, and `webpackFinal`
- Add description for `core.channelOptions`
- Remove the `previewMainTemplate` page
- Document possible shortcomings of some `stories` configurations
- Document limitations of some `typescript` options

jonniebigodes commented on docs/snippets/common/main-config-framework.js.mdx:
```diff
@@ -0,0 +1,14 @@
+export default {
+  framework: {
+    // Replace react-vite with the framework you are using (e.g., react-webpack5)
+    name: '@storybook/react-vite',
```
Per our conversation and to align with the other examples, replace `@storybook/react-vite` with the framework agnostic text. Similar to the the TS variation

kylegach replied:
I had left this one using `react-vite` on purpose, because the options demonstrated only apply to React projects. The comment is unique to this snippet too, saying to replace `react-vite` with `react-webpack5`, to keep it React-specific.

Make sense? Bad idea?

jonniebigodes replied:
Thanks for the clarification on this. appreciate it

jonniebigodes reviewed (changes requested):
One small item to address, and this should be good to go. Let me know and we'll go from there.

jonniebigodes reviewed (changes requested):
One small item to address, and this should be good to go. Let me know and we'll go from there.
